interfaces: Expose settings.json methods to GUI

ryanofsky opened the pull request:
Add `interfaces::Node` `updateSetting`, `forceSetting`, `resetSettings`, `isSettingIgnored`, and `getPersistentSetting` methods so GUI is able to manipulate `settings.json` file and use and modify node settings.

(Originally this PR also contained GUI changes to unify bitcoin-qt and bitcoind persistent settings and call these methods, but the GUI commits have been dropped from this PR and moved to bitcoin-core/gui/pull/602)

ryanofsky committed:
settings: Add update/getPersistent/isIgnored methods

Add interfaces::Node methods to give GUI finer grained control over
settings.json file. Update method is used to write settings to the file,
getPersistent and isIgnored methods are used to find out about settings
file and command line option interactions.

ryanofsky committed:
init: Remove Shutdown() node.args reset

This commit removes the `node.args = nullptr` assignment in the Shutdown()
function.

Clearing node.args there never made sense because it made the
Shutdown() function not idempotent, making it fragile and causing issues like
https://github.com/bitcoin/bitcoin/issues/23186.

The assignment also causes segfaults in GUI unit tests when a new
node().initParameterInteraction() call is added in OptionsModel to apply to Qt
settings (happens because AppTests calls Shutdown() which sets node.args to
null, and OptionTests runs after AppTests and then needs node.args not to be
null.)

jonasschnelli commented:
Concept ACK

ryanofsky commented:
Travis compile error in https://travis-ci.org/bitcoin/bitcoin/jobs/620388014#L3988 seems to be due to a clang bug with macros and raw strings: https://reviews.llvm.org/D39279

```
qt/test/optiontests.cpp:56:100: warning: missing terminating '"' character [-Winvalid-pp-token]
)", "std::string(std::istreambuf_iterator<char>(file), std::istreambuf_iterator<char>()).c_str()", "R\"({
                                                                                                   ^
qt/test/optiontests.cpp:56:100: error: expected expression
```

Maybe can work around by adding parentheses, otherwise could declare temp variable.

---

Other travis error https://travis-ci.org/bitcoin/bitcoin/jobs/620388019#L2929 seems like a qsettings int value gets stringified on mac (doesn't happen on linux):

Actual: `"dbcache": "600",`
Expected: `"dbcache": 600,`

It would probably be good to fix this by adding an explicit type argument to [`ToSetting`](https://github.com/ryanofsky/bitcoin/blob/955f7c00a161d6477172fa9760d1b2b5626d1761/src/qt/optionsmodel.cpp#L31-L40) instead of trying to guess the setting type from the variant type

ryanofsky commented:
Rebased 17fcdf51ab570eb55b46a7c44a2a631739462612 -> ec7cc41c00ca93a8bbf92ec13101358f7472cd84 ([`pr/qtset.22`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.22) -> [`pr/qtset.23`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.23), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/qtset.22-rebase..pr/qtset.23)) due to conflict with bitcoin-core/gui#416

prayank23 commented:
Concept ACK

MarcoFalke commented on src/qt/optionsmodel.cpp:
```diff
@@ -606,4 +625,41 @@ void OptionsModel::checkAndMigrate()
+    migrate_setting(MapPortUPnP, "fUseUPnP", "upnp");
+    migrate_setting(MapPortNatpmp, "fUseNatpmp", "natpmp");
+    migrate_setting(Listen, "fListen", "listen");
+    migrate_setting(Server, "server", "server");
```
Is this needed, given that no version was released with this setting?

ryanofsky replied:
> Is this needed, given that no version was released with this setting?

Not strictly needed, but it would be strange to remove this because removing it would treat this setting differently than all the other settings, it causes no harm, it is one line of code, removing it would introduce strange behaviors trying to revert and test this PR, and removing it would introduce a bug in the PR if there is any release before this PR is merged.

prayank23 commented:
Tried on Fedora and it doesn't work as expected.

```
[prayank@fedora bin]$ ./bitcoin-qt
QSocketNotifier: Can only be used with threads started with QThread
Segmentation fault (core dumped)
```

I have `txindex=1` in bitcoin.conf, not a pruned node which works fine with `bitcoind`. It created a settings file on first launch of `bitcoin-qt` which failed with error that `txindex` doesn't work with pruning. Tried running with `prune=0` and it still doesn't work.

I expect `bitcoin-qt` should use prune=0 by default if nothing is mentioned in config and also fix this segfault error.

ryanofsky reviewed (commented):
Updated 4acc7fcbbcc63a25ac1e5368141dc9612a73c0d1 -> 6e1c1ea20b8cf0bb6c24c81cda92e79dc14c31d5 ([`pr/qtset.24`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.24) -> [`pr/qtset.25`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.25), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/qtset.24..pr/qtset.25)) to fix segfault caused by bad rebase.

---

> Tried on Fedora and it doesn't work as expected.

Thanks, segfault was caused by a conflict in one of the rebases and should be fixed in the newest update of this PR.

> I expect `bitcoin-qt` should use prune=0 by default if nothing is mentioned in config and also fix this segfault error.

The segfault error shouldn't be related to the prune or txindex settings, and I couldn't reproduce any expected behaviors experimenting with these settings myself. I think the segfault should disappear if you try the updated version of the PR. And if you continue seeing prune/txindex errors, it'd be helpful to see contents of `~/.config/Bitcoin/Bitcoin-Qt.conf` and `~/.bitcoin/settings.json` files. (if you are using regtest the files are `~/.config/Bitcoin/Bitcoin-Qt-regtest.conf` and `~/.bitcoin/regtest/settings.json`)

---

CI failure https://cirrus-ci.com/task/4987964057976832 `Insufficient funds` in `rpc_fundrawtransaction.py` was probably spurious and might be related to #22949.

ryanofsky commented:
Rebased 6e1c1ea20b8cf0bb6c24c81cda92e79dc14c31d5 -> a4111bdc448af8f2b5a7126164fb7e5bb6c7f2b0 ([`pr/qtset.25`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.25) -> [`pr/qtset.26`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.26), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/qtset.25-rebase..pr/qtset.26)) due to conflict with #22951

ryanofsky commented:
Thanks for the testing! In summary, I think all behavior described is expected (and improves on current behavior, even if could be improved more later).

> 1. `bitcoin-qt` had prune option unchecked in GUI. Relaunched `bitcoin-qt` with a data directory that has pruned node. Everything works fine and it automatically checked the prune option. Although GB mentioned in GUI was incorrect. I had saved prune=1024 in `bitcoin.conf` and GUI had 2 GB.

The GUI prune setting is a whole number of GB so it's easier to change and understand than if it were MiB, and to avoid mistakes editing large numbers. 1024MiB is rounded up to 2GB in the display, even though 1024MiB should still be the effective pruning size until the user changes it again. The idea behind rounding up is that it would be bad if node used more disk space than the displayed amount, but ok if it's using less space than the displayed amount.

> 2. `bitcoin-qt` had prune option checked in GUI from 1. Used data directory that has all blocks saved on disk. Everything works fine and it automatically unchecked the prune option in GUI.

Nice test. It's good to have confirmation that `bitcoin.conf` still takes precedence over `QSettings` so upgrading does not cause changed settings to be used. (Implementation note: the precedence for settings sources is `settings.json` > `bitcoin.conf` > `QSettings`)

> 3. I wanted to check what happens if qt and bitcoin.conf are different. Restarted `bitcoin-qt` with data directory used in 1. Unchecked the prune option manually in GUI. This saved `"prune": 0` in `settings.json`. While `bitcoin.conf` for 1 still has `prune=1024`, started bitcoin-qt and I see this error: [Do you want to rebuild the block database now?]

Ideally this would be shown as a question not an error, but the benefit of this PR is that when you uncheck the pruning option it now actually disables pruning. Disabling pruning does require reindexing, which is why it asks to reindex.

Speaking more generally, the idea behind settings precedence is that `settings.json` is intended to hold live, dynamic settings chosen by the user, and `bitcoin.conf` is supposed to hold static settings chosen by the packager or installer (who can disable live settings with `nosettings=1`). The cases where different settings sources provide different values are intended to be well defined, but also rare. There should be few use cases that require controlling the same setting from different places. In the most common cases, GUI users should not need any `bitcoin.conf` file at all. GUI users should have better options than creating a text file, learning a configuration syntax, and restarting their node to apply settings. Also, ideally more settings will be able to be applied live without restarts, and there will be a `config` RPC method to get and set arbitrary `settings.json` settings.

> 4. Saved below things in `settings.json` for non-pruned node:
>    ```
>    {
>        "prune": 1024,
>        "wallet": [
>        ]
>       }
>    ```
>
>    `bitcoin.conf` had `txindex=1`
> 
> and this is the error I get: [Prune mode is incompatible with -txindex]

This is another good test, and I think it is straightforward that if you enable pruning and enable txindex you will see this error. Ideally this could show options for disabling pruning or disabling txindex, though.

> 5. There is one more case which I wanted to reproduce but couldn't and also don't want to risk deleting my blocks for non-pruned node which is explained in [Pruning should be based on bitcoin.conf bitcoin-core/gui#245](https://github.com/bitcoin-core/gui/issues/245)

I don't think there should be any direct interaction between this PR and what's described in [#245](https://github.com/bitcoin-core/gui/issues/245). With this PR the node will still prune if you enable pruning in the GUI, and still not prune if you disable pruning in the GUI. The other issue [#245](https://github.com/bitcoin-core/gui/issues/245) seems like it is requesting that the node avoids pruning even if pruning was requested, which does not make sense to me even though I agree we should have safeguards to prevent pruning from happening unintentionally. If you can describe in more detail how [#245](https://github.com/bitcoin-core/gui/issues/245) happened (how pruning got enabled in the GUI), maybe we could add some safeguard to prevent this.

ryanofsky commented:
> If you can describe in more detail how [#245](https://github.com/bitcoin-core/gui/issues/245) happened (how pruning got enabled in the GUI), maybe we could add some safeguard to prevent this.

Maybe the GUI should have some an extra prompt at startup pruning is changing from disabled to enabled and blocks exist. In the opposite case we already have "You need to rebuild the database using -reindex to go back to unpruned mode.  [...] Do you want to rebuild the block database now?". In this case GUI could prompt "Pruning setting has changed from disabled to enabled, which may discard previously downloaded blocks. Are you sure you want to enable pruning?" Buttons could be [Enable pruning] [Disable pruning] [Abort]." If this is a good idea to implement, probably it should be a separate GUI-only PR.

ryanofsky commented:
Rebased a4111bdc448af8f2b5a7126164fb7e5bb6c7f2b0 -> 32ef6cfa521b8a0615de3259d7b81383608ab8f3 ([`pr/qtset.26`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.26) -> [`pr/qtset.27`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.27), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/qtset.26-rebase..pr/qtset.27)) due to conflict with #23004

prayank23 commented:
reACK https://github.com/bitcoin/bitcoin/commit/32ef6cfa521b8a0615de3259d7b81383608ab8f3

One user has some issues with prune settings today: https://bitcoin.stackexchange.com/questions/110994/bitcoin-core-just-deleted-my-350-gb-of-blockchain-data/

ryanofsky commented:
Rebased 32ef6cfa521b8a0615de3259d7b81383608ab8f3 -> 91d3b48508440a16818ee9606bf5bbdca82edf8a ([`pr/qtset.27`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.27) -> [`pr/qtset.28`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.28), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/qtset.27-rebase..pr/qtset.28)) due to conflict with bitcoin-core/gui#441

prayank23 reviewed (approved):
reACK https://github.com/bitcoin/bitcoin/pull/15936/commits/91d3b48508440a16818ee9606bf5bbdca82edf8a

Hoping this gets reviewed soon and merged.

hebasto commented:
Concept ACK.

hebasto reviewed (commented):
Approach ACK 91d3b48508440a16818ee9606bf5bbdca82edf8a.

In here
```
$ cat settings.json 
{
    "fUseNatpmp": false,
    "listen": false,
    "upnp": false,
    "wallet": [
     ""
    ]
   }
```
I was expecting to see `natpmp` instead of `fUseNatpmp`.

ryanofsky commented:
> I was expecting to see `natpmp` instead of `fUseNatpmp`.

Thanks for the review, and good catch. I think this was a bug introduced during a rebase. I also fixed a similar problem with `-lang` flag / `language` setting, and split off a separate commit introducing the OptionsModel getOption/setOption methods to make the individual settings changes easier to see.

Updated 91d3b48508440a16818ee9606bf5bbdca82edf8a -> 79a0399c696b8212c11c868e1f45f9887aa71993 ([`pr/qtset.28`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.28) -> [`pr/qtset.29`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.29), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/qtset.28..pr/qtset.29)) fixing natpmp and language bugs and splitting the commit

Rspigler commented:
Testing 79a0399c696b8212c11c868e1f45f9887aa71993

Opening `settings.json` with a fresh Bitcoin-qt (I have some wallets in the directory).

> 
> {
>     "wallet": [
>      "1",
>      "2",
>      "Config",
>      "1.6.22"
>     ]
>    }

Open Bitcoin-qt, through the GUI set prune to 4GB.  Close Bitcoin-qt.  Open `settings.json` again:

> {
>     "prune": 3814,
>     "upnp": false,
>     "wallet": [
>      "1",
>      "2",
>      "Config",
>      "1.6.22"
>     ]
>    }

UpnP is included because it is by default unselected in the GUI? I think that makes sense.

I create a `bitcoin.conf` and set `blockfilterindex=1`.  I get an alert:

> 
Error: basic block filter index best block of the index goes beyond pruned data. Please disable the index or reindex (which will download the whole blockchain again)

This makes sense.  So I run `bitcoin-qt -reindex`

However, I open `settings.json`, and it doesn't show blockfilterindex:

> 
> {
>     "prune": 3814,
>     "upnp": false,
>     "wallet": [
>      "1",
>      "2",
>      "Config",
>      "1.6.22"
>     ]
>    }

And I open `bitcoin.conf` and it doesn't show any other settings besides `blockfilterindex=1`

hebasto reviewed (commented):
Tested 79a0399c696b8212c11c868e1f45f9887aa71993.

Before running `bitcoin-qt`:
```
$ cat settings.json 
{
    "dbcache": 444,
    "lang": "en_GB",
    "listen": false,
    "natpmp": true,
    "onion": "127.0.0.1:9050",
    "par": -1,
    "proxy": "127.0.0.1:9050",
    "prune": 3814,
    "upnp": true,
    "wallet": [
     ""
    ]
   }
```
Please note `"natpmp": true` and `"upnp": true`.

Running `bitcoin-qt`, then Options shows:

![DeepinScreenshot_select-area_20220114115424](https://user-images.githubusercontent.com/32963518/149496051-8da8e430-d83d-4c93-847f-0539a0462104.png)

ryanofsky commented:
Rspigler, all of that behavior except the upnp part is working as designed. The goal of the PR is for bitcoin-qt and bitcoind to share the same dynamic settings, and to store those settings in the settings.json file instead of QT settings storage. The software will _never_ write or change bitcoin.conf because bitcoin.conf is for static settings, rather than dynamic ones. The ultimate goal (which is not fully realized here, but would be with a `bitcoin-cli settings` command) is that nobody should ever need to create or edit a `bitcoin.conf` file to change any setting. `bitcoin.conf` would be a way for distributions and packages to set configuration defaults, and for advanced users to be able switch between complicated configurations, or to track their configurations in git, or generate their configurations with deployment tools like docker, ansible, or nix. But everyday users would not know or care about `bitcoin.conf` and would just use the gui or command line to change settings dynamically (CLI would look like `bitcoin-cli settings set prune 4000`) with immediate feedback about invalid values, and not have to worry about creating a text file, or dealing with a configuration syntax, or stopping and restarting bitcoin to make sure the file is parsed and their settings are valid.

Hebasto, it appears there's another bug with natpmp/upnp settings, and I think I see what it is: the `OptionsModel::getOption` method is still returning `settings.value("fUseUPnP")` and `settings.value("fUseNatpmp")` for these settings instead of `ToQVariant(node().getPersistentSetting("upnp"), DEFAULT_UPNP)` and `ToQVariant(node().getPersistentSetting("natpmp"), DEFAULT_NATPMP)`. There may be additional bugs around these settings as well. The implementation of the nat settings changed a lot after this PR was opened, and while I rebased it and fixed explicit conflicts, I never really went through the implementation updating all the upnp/natpmp references that needed to be updated or tested this feature.

ClaraBara22 commented:
I have removed the VPN on my device. I hope it helps.

ryanofsky reviewed (commented):
Rebased 79a0399c696b8212c11c868e1f45f9887aa71993 -> 7eb8fa3daf702d4f609ae0532dcfb61e4d26a173 ([`pr/qtset.29`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.29) -> [`pr/qtset.30`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.30), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/qtset.29-rebase..pr/qtset.30)) due to conflicts with #24498 and ##20744. Also fixed upnp issue discussed above, and split PR into smaller commits.

ryanofsky commented:
Updated 7eb8fa3daf702d4f609ae0532dcfb61e4d26a173 -> 52438b033718afae375cea86ca349db3a878aef4 ([`pr/qtset.30`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.30) -> [`pr/qtset.31`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.31), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/qtset.30..pr/qtset.31)) switching some default literals to constants

hebasto commented:
FWIW, this PR (52438b033718afae375cea86ca349db3a878aef4) fixes the bitcoin-core/gui#567 regression.

jonatack commented:
Concept ACK

vasild commented:
Concept ACK

ryanofsky commented:
> FWIW, this PR ([52438b0](https://github.com/bitcoin/bitcoin/commit/52438b033718afae375cea86ca349db3a878aef4)) fixes the [bitcoin-core/gui#567](https://github.com/bitcoin-core/gui/issues/567) regression.

I could be wrong, but I'm guessing this PR only partially fixes https://github.com/bitcoin-core/gui/issues/567. It should prevent the bug if someone disables listening with this PR applied, and then restarts bitcoin-qt. But I'm guessing if they turn off listening without this PR, then restart with this PR applied, the same "Cannot set -listen=0 together with -listenonion=1" error will happen, because migrate_settings code apply the setting too late after the InitParameterInteraction code has already run, but before the AppInitParameterInteraction code runs and triggers the error. But the bug should be more temporary and go away after the next restart after the setting is migrated.

ryanofsky commented:
Rebased 52438b033718afae375cea86ca349db3a878aef4 -> b81c22a2880a797ea86d9b793c58aa49c67b9cda ([`pr/qtset.31`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.31) -> [`pr/qtset.32`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.32), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/qtset.31-rebase..pr/qtset.32)) due to conflict with bitcoin-core/gui#568. Also cleaned up messy previous conflict resolution with #17696. No change in behavior except for a minor bugfix when the -prune command line option is given and the intro dialog is shown. In previous version of this PR, the intro dialog prune value would only be written to settings.json and not take precedence over the command line value. Now it takes precedence over the command line value.

ryanofsky commented:
Rebased b81c22a2880a797ea86d9b793c58aa49c67b9cda -> 8d50aba09faafa2329751a30d5ce6f63afdc52d0 ([`pr/qtset.32`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.32) -> [`pr/qtset.33`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.33), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/qtset.32-rebase..pr/qtset.33)) fixing conflict and test failure after bitcoin-core/gui#569

hebasto commented:
https://api.cirrus-ci.com/v1/task/6080882288099328/logs/ci.log:
```
FAIL: qt/test/test_bitcoin-qt
=============================

QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/tmp/runtime-root'
********* Start testing of AppTests *********
Config: Using QtTest library 5.15.3, Qt 5.15.3 (x86_64-little_endian-lp64 static release build; by Clang 8.0.0 (tags/RELEASE_800/final)), ubuntu 18.04
PASS   : AppTests::initTestCase()
QINFO  : AppTests::appTests() Backing up GUI settings to "/tmp/test_common_Bitcoin Core/95f88467c15a125c7116b7e203b7386cf60ef5f9ce394165bee770ab2655162d/regtest/guisettings.ini.bak"
QDEBUG : AppTests::appTests() requestInitialize : Requesting initialize
QDEBUG : AppTests::appTests() Running initialization in thread
QDEBUG : AppTests::appTests() initializeResult : Initialization result:  true
QINFO  : AppTests::appTests() Platform customization: "other"
QWARN  : AppTests::appTests() This plugin does not support propagateSizeHints()
QWARN  : AppTests::appTests() This plugin does not support raise()
QWARN  : AppTests::appTests() This plugin does not support grabbing the keyboard
QWARN  : AppTests::appTests() This plugin does not support propagateSizeHints()
QDEBUG : AppTests::appTests() requestShutdown : Requesting shutdown
QDEBUG : AppTests::appTests() Running Shutdown in thread
QDEBUG : AppTests::appTests() Shutdown finished
PASS   : AppTests::appTests()
PASS   : AppTests::cleanupTestCase()
Totals: 3 passed, 0 failed, 0 skipped, 0 blacklisted, 522ms
********* Finished testing of AppTests *********
********* Start testing of OptionTests *********
Config: Using QtTest library 5.15.3, Qt 5.15.3 (x86_64-little_endian-lp64 static release build; by Clang 8.0.0 (tags/RELEASE_800/final)), ubuntu 18.04
PASS   : OptionTests::initTestCase()
FAIL!  : OptionTests::migrateSettings() Caught unhandled exception
   Loc: [qtestcase.cpp(1939)]
Totals: 1 passed, 1 failed, 0 skipped, 0 blacklisted, 7ms
********* Finished testing of OptionTests *********
terminating with uncaught exception of type std::__1::__fs::filesystem::filesystem_error: filesystem error: in create_directory: Not a directory [/root/.bitcoin/regtest]
FAIL qt/test/test_bitcoin-qt (exit status: 134)
```

ryanofsky commented:
Thanks! I think the errors should be fixed by the last push.

Updated 8d50aba09faafa2329751a30d5ce6f63afdc52d0 -> 73d7e74aa3201158a7f33daf07fc23a370d2b4b9 ([`pr/qtset.33`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.33) -> [`pr/qtset.34`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.34), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/qtset.33..pr/qtset.34)) to fix create_directory errors after OptionsTest caused by OptionsTest clearing the -datadir option set by BasicTestingSetup (https://cirrus-ci.com/task/4954982381256704?logs=ci#L3402, https://cirrus-ci.com/task/5377194846322688?logs=ci#L2727, https://cirrus-ci.com/task/4814244892901376?logs=ci#L3122, https://cirrus-ci.com/task/5236457357967360?logs=ci#L4040, https://cirrus-ci.com/task/4673507404546048?logs=ci#L3145, https://cirrus-ci.com/task/4954982381256704?logs=ci#L3402, https://cirrus-ci.com/task/6080882288099328?logs=ci#L2910)

vasild reviewed (commented):
Posting mid-review, still need to review the last commit and test. Looks good so far.

ryanofsky reviewed (commented):
Updated 73d7e74aa3201158a7f33daf07fc23a370d2b4b9 -> 88ead7707dea2e31ce1245daee7d23be3f72e041 ([`pr/qtset.34`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.34) -> [`pr/qtset.35`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.35), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/qtset.34..pr/qtset.35)) with fix for macos test failure. The failure just happened because optionsTest code dependended on appTests testing setup, and appTests are skipped by default on mac. Problem could be reproduced on linux skipping appTests there.

vasild commented on src/qt/optionsmodel.cpp:
```diff
@@ -19,18 +19,103 @@
+    return LocaleIndependentAtoi<int>(value.get_str());
+}
+
+//! Get pruning enabled value to show in GUI from bitcoin -prune setting.
+static bool PruneEnabled(const util::SettingsValue& prune_setting)
+{
+    // -prune=1 setting is manual pruning mode, so disabled for purposes of the gui
+    return ToInt(prune_setting) > 1;
```
This, in isolation, is (wrongly) ignoring the fact that values in [2, 550) are not allowed (`MIN_DISK_SPACE_FOR_BLOCK_FILES` -- 550 MiB). `init.cpp` would refuse to start with such a value.

In context, this is only called with values entered in the GUI which is accepting integer GB only, so it is impossible to enter e.g. 100 MiB. Or it could come from `settings.json` where the value is stored in MiB and the user may have edited it (they are not supposed to). I think it is ok because even if <550 is entered in `settings.json` the startup would gracefully fail with the proper message from `init.cpp`.

vasild replied:
Can be marked as resolved.

vasild reviewed (commented):
Review still ongoing, posting two minor comments coz I am afraid github may lose them, feel free to ignore.

This would be easier to review if the last commit `Unify bitcoin-qt and bitcoind persistent settings` is split so that there is one commit per config option migrated from `~/.config/Bitcoin/Bitcoin-Qt.conf` to `settings.json`.

vasild commented on src/qt/optionsmodel.cpp:
```diff
@@ -19,18 +19,93 @@
+    if (value.isNull()) return fallback;
+    if (value.isBool()) return value.get_bool();
+    if (value.isNum()) return value.get_int();
+    return QString::fromStdString(value.get_str());
```
Ok, leave it as is then.

I mentioned it because this function has a return type `QVariant` and now it returns `QString`. This, I guess, implicitly calls the `QVariant(const QString &string)` contructor. Where is `QString::fromStdString()` documented? Is the constructor `QVariant(const QString &string)` documented too? I do not see any comments in Qt headers.

vasild commented on src/qt/optionsmodel.cpp:
```diff
@@ -265,322 +256,337 @@ int OptionsModel::rowCount(const QModelIndex & parent) const
-// read QSettings values and return them
-QVariant OptionsModel::data(const QModelIndex & index, int role) const
+// write QSettings values
+bool OptionsModel::setData(const QModelIndex & index, const QVariant & value, int role)
```
ok

vasild commented on src/qt/bitcoin.cpp:
```diff
@@ -259,7 +259,7 @@ void BitcoinApplication::createPaymentServer()
 
 void BitcoinApplication::createOptionsModel(bool resetSettings)
 {
-    optionsModel = new OptionsModel(this, resetSettings);
+    optionsModel = new OptionsModel(node(), this, resetSettings);
```
Ok, leave it as is for the purposes of this PR.

There was some subtle bug recently caused by some not-obvious dependency during startup, but the difference with this code here is that here we have an `assert()`, so it will not remain unnoticed.

vasild commented on src/qt/bitcoin.cpp:
```diff
@@ -632,6 +632,7 @@ int GuiMain(int argc, char* argv[])
     app.parameterSetup();
     GUIUtil::LogQtInfo();
     // Load GUI settings from QSettings
```
ok

vasild replied:
Can be marked as resolved.

vasild reviewed (commented):
The ip address and port of the proxy would be lost if the user unchecks the "Connect through proxy..." in the network settings. Further, `bitcoin-qt` refuses to start with the error message:

```
Error: No proxy server specified. Use -proxy=<ip> or -proxy=<ip:port>.
```

To reproduce:

1. In the "GUI settings -> options -> network" click "Connect through SOCKS5 proxy" and input some values for "Proxy IP" and "Port", l used `127.0.0.2` and `19050`. Shut down.

2. Confirm that `~/.bitcoin/settings.json` contains `"proxy": "127.0.0.2:19050"`. The use-proxy checkbox is not saved explicitly. Rather the presence of `"proxy"` implies that it is checked.

3. Start and uncheck the "Connect through SOCKS5 proxy" box in the GUI. Notice that the address and port of the proxy are still displayed but greyed out / disabled. This is ok. In order to re-enable the proxy the user has to only check the box later and the previous addr/port from before are already filled in. Shut down.

4. Notice that `~/.bitcoin/settings.json` contains `"proxy": ""`. At this point the user's values for proxy address and port have been lost.

5. Try to start `bitcoin-qt`, it refuses to start with an error `Error: No proxy server specified`. This is like trying to start `bitcoind` with `-proxy=""`.

I guess it would be better to explicitly save a "use proxy" boolean in `settings.json` that corresponds to the "Connect through SOCKS5 proxy" checkbox. This way the user can temporarily disable the proxy and re-enable it later, through restart, without having to re-enter its address and port. Also, if that "use proxy" boolean is `false` then we should avoid setting the `proxy` config which should allow `bitcoin-qt` to start.

In this case my comments below about removing `is_set` from `ProxySettings` below should be ignored.

ryanofsky reviewed (commented):
Thanks @vasild for the review and bug report and very helpful suggestions!

Updated 88ead7707dea2e31ce1245daee7d23be3f72e041 -> 1a8ac5bb1baf555003580db96d619f4a9c23089a ([`pr/qtset.35`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.35) -> [`pr/qtset.36`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.36), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/qtset.35..pr/qtset.36)) with suggested changes

re: https://github.com/bitcoin/bitcoin/pull/15936#pullrequestreview-936298972

> This would be easier to review if the last commit `Unify bitcoin-qt and bitcoind persistent settings` is split so that there is one commit per config option migrated from `~/.config/Bitcoin/Bitcoin-Qt.conf` to `settings.json`.

Thanks, I agree and split this up into multiple commits.

re: https://github.com/bitcoin/bitcoin/pull/15936#pullrequestreview-937685759

> The ip address and port of the proxy would be lost if the user unchecks the "Connect through proxy..." in the network settings. Further, `bitcoin-qt` refuses to start with the error message:

It seems ok to me (and perhaps better) for the old proxy string to be cleared if the proxy is disabled and bitcoin-qt is restarted. I could think of various approaches to saving the unused proxy value, though, if it is important to change this.

The "No proxy server specified" error message is definitely a bug. I could think of various ways to fix it, but the approach I like best is #24830, because it tones down the overeager error message more generally and doesn't require changes here.

ryanofsky commented:
I updated this to drop all the ToSetting/ToQVariant/ToQString/ToInt functions. These were supposed to help make optionsmodel code simpler, but I think in practice I think they were only obfuscating it. They were also duplicating type conversion logic in ArgsManager, which I think is just better to expose directly.

Updated 1a8ac5bb1baf555003580db96d619f4a9c23089a -> 9b78c241e5e3af50eec732fb044183331fd0c756 ([`pr/qtset.36`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.36) -> [`pr/qtset.37`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.37), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/qtset.36..pr/qtset.37)) fixing missing univalue include.
Updated 9b78c241e5e3af50eec732fb044183331fd0c756 -> 7966254822e179798d6a375ac3183321ee541733 ([`pr/qtset.37`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.37) -> [`pr/qtset.38`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.38), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/qtset.37..pr/qtset.38)) dropping setting conversion functions and cleaning up implementation and tests in other small ways.

vasild commented on src/qt/optionsmodel.cpp:
```diff
@@ -637,4 +617,49 @@ void OptionsModel::checkAndMigrate()
         settings.setValue("addrSeparateProxyTor", GetDefaultProxyAddress());
     }
+
+    // Migrate and delete legacy GUI settings that have now moved to <datadir>/settings.json.
```
This migration may silently drop configs from `.config/Bitcoin/Bitcoin-Qt.conf` with unexpected effects. Before this PR, values in `Bitcoin-Qt.conf` were overriding values from `bitcoin.conf`. So, for example, if one had:

`bitcoin.conf`: `listen=1`
`Bitcoin-Qt.conf `: `fListen=false`

then `bitcoin-qt` would not be listening and "Allow incoming connections" will be unchecked in the GUI (correct).

However, once the migration happens, it would leave the configs in this state:

`bitcoin.conf`: `listen=1`
`Bitcoin-Qt.conf `: (no fListen is present)

which will cause `bitcoin-qt` to start listening, silently, when it was not listening before.

vasild replied:
Thanks for the elaborate explanation! I was wrong, sorry for the noise.

Can be marked as resolved.

vasild commented on src/qt/optionsmodel.cpp:
```diff
@@ -19,18 +19,59 @@
+}
+
+struct ProxySetting {
+    bool is_set;
```
_(continuing the proxy [discussion](https://github.com/bitcoin/bitcoin/pull/15936#pullrequestreview-937685759) here so that it is in its own thread, not in the main PR comments and so that it can be closed when resolved eventually)_

This applies to the options that have a user-input value and an additional checkbox to disable them: `prune`, `proxy` and `onion`. Naturally there are two things per option - the value typed by the user and the boolean checkbox (denoting enabled/disabled).

This PR uses a dummy value in `settings.json` to denote the "disabled" state: `0` for `prune` and `""` for `proxy`. The problem with this approach is that this loses the value typed by the user. This is surprising and annoying (having to re-enter it after temporary disabling it). Also, it is not consistent behavior:

* closing the config window and re-opening it => the values are still there, greyed out
* restarting `bitcoin-qt` => the values are deleted

Firefox for example has a proxy-config and it acts in a consistent way - the values, when disabled, are preserved (greyed out) even after restart. IMO this is [less surprising](https://en.wikipedia.org/wiki/Principle_of_least_astonishment).

I would suggest to save the "use prune", "use proxy" and "use proxy for tor" as explicit boolean entities in `settings.json` and preserve the user-entered values in `prune`, `proxy` and `onion`.

Actually, for `prune` the behavior is even a bit more confusing:
* enter 5 (some arbitrary value)
* disable the checkbox (5 remains in the box, greyed out)
* restart
* enable the checkbox
* surprise! the value is not 5, but 2

vasild replied:
> I'll try to implement a patch which implements the previous behavior you seem to like.

Hmm, I shouldn't be forcing my view here. I could be nuts and totally disconnected from realit:space_invader: Maybe at this point it makes sense to poke other reviewers for opinion. Anyone?

> I think it's ok to clear pruning or proxy values after a users disables pruning or proxying and then restarts.

I think it would be less surprising and more obvious what is going on if we clear the values after the user disables the option. Like, immediately after the click on the checkbox. Then the behavior will be consistent between "close & re-open the config dialog box" and "close & re-open the app".

To summarize, there are 3 approaches for when an option is disabled (sorted in order of my preference):

1. Preserve the value (greyed out) through config dialog re-open and through app restart. To achieve this, store an explicit `use_proxy: false` in `settings.json` and keep the proxy address in `proxy: "..."`. This means that a config which `bitcoind` does not recognize will appear in `settings.json` which will require a tweak here: https://github.com/bitcoin/bitcoin/blob/b297b945f7610772434817181ad12067b2832565/src/util/system.cpp#L569

2. Delete the value immediately after the checkbox is unchecked. This will make it obvious that the value is gone.

3. Preserve the value (greyed out) through config dialog re-open, but delete it through an app restart. This is how this PR works as of now (99b4ce266f62afc8a37575527d82ac6048f50f75).

vasild reviewed (commented):
Reviewed all as of 7966254822e179798d6a375ac3183321ee541733. Thanks for splitting the last commit, much easier to review now. Hopefully it will be easier for other reviewers too.

ryanofsky reviewed (commented):
Thanks again vasild for the thorough review and testing! I made some updates and fixes in response to you comments, and also replied to everything below.

Rebased 7966254822e179798d6a375ac3183321ee541733 -> 99b4ce266f62afc8a37575527d82ac6048f50f75 ([`pr/qtset.38`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.38) -> [`pr/qtset.39`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.39), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/qtset.38-rebase..pr/qtset.39)) due to conflict with bitcoin-core/gui#556, also fixing uncaught startup exception on setting.json pointed out by vasild, and adding an options id to setting name helper function to reduce code duplication.

vasild commented on src/qt/optionsmodel.cpp:
```diff
@@ -273,322 +252,345 @@ int OptionsModel::rowCount(const QModelIndex & parent) const
+
+    // default proxy
+    case ProxyUse:
+        return ParseProxyString(SettingToString(setting(), "")).is_set;
```
This diff is now in commit ce530d566f `Migrate -prune setting from QSettings to settings.json`, but it has to be amended to commit dd3d146f72 `Migrate -proxy and -onion settings from QSettings to settings.json`

```diff
--- a/src/qt/optionsmodel.cpp
+++ b/src/qt/optionsmodel.cpp
@@ -338,21 +364,21 @@ QVariant OptionsModel::getOption(OptionID option) const
 #endif // USE_NATPMP
     case MinimizeOnClose:
         return fMinimizeOnClose;
 
     // default proxy
     case ProxyUse:
-        return ParseProxyString(SettingToString(node().getPersistentSetting("proxy"), "")).is_set;
+        return ParseProxyString(SettingToString(setting(), "")).is_set;
     case ProxyIP:
         return m_proxy_ip;
     case ProxyPort:
         return m_proxy_port;
 
     // separate Tor proxy
     case ProxyUseTor:
-        return ParseProxyString(SettingToString(node().getPersistentSetting("onion"), "")).is_set;
+        return ParseProxyString(SettingToString(setting(), "")).is_set;
     case ProxyIPTor:
         return m_onion_ip;
     case ProxyPortTor:
         return m_onion_port;
 
 #ifdef ENABLE_WALLET
@@ -430,56 +456,56 @@ bool OptionsModel::setOption(OptionID option, const QVariant& value)
         settings.setValue("fMinimizeOnClose", fMinimizeOnClose);
         break;
 
     // default proxy
     case ProxyUse:
         if (changed()) {
-            node().updateSetting("proxy", ProxyString(value.toBool(), m_proxy_ip, m_proxy_port));
+            update(ProxyString(value.toBool(), m_proxy_ip, m_proxy_port));
             setRestartRequired(true);
         }
         break;
     case ProxyIP:
         if (changed()) {
             m_proxy_ip = value.toString();
             if (getOption(ProxyUse).toBool()) {
-                node().updateSetting("proxy", ProxyString(true, m_proxy_ip, m_proxy_port));
+                update(ProxyString(true, m_proxy_ip, m_proxy_port));
                 setRestartRequired(true);
             }
         }
         break;
     case ProxyPort:
         if (changed()) {
             m_proxy_port = value.toString();
             if (getOption(ProxyUse).toBool()) {
-                node().updateSetting("proxy", ProxyString(true, m_proxy_ip, m_proxy_port));
+                update(ProxyString(true, m_proxy_ip, m_proxy_port));
                 setRestartRequired(true);
             }
         }
         break;
 
     // separate Tor proxy
     case ProxyUseTor:
         if (changed()) {
-            node().updateSetting("onion", ProxyString(value.toBool(), m_onion_ip, m_onion_port));
+            update(ProxyString(value.toBool(), m_onion_ip, m_onion_port));
             setRestartRequired(true);
         }
         break;
     case ProxyIPTor:
         if (changed()) {
             m_onion_ip = value.toString();
             if (getOption(ProxyUseTor).toBool()) {
-                node().updateSetting("onion", ProxyString(true, m_onion_ip, m_onion_port));
+                update(ProxyString(true, m_onion_ip, m_onion_port));
                 setRestartRequired(true);
             }
         }
         break;
     case ProxyPortTor:
         if (changed()) {
             m_onion_port = value.toString();
             if (getOption(ProxyUseTor).toBool()) {
-                node().updateSetting("onion", ProxyString(true, m_onion_ip, m_onion_port));
+                update(ProxyString(true, m_onion_ip, m_onion_port));
                 setRestartRequired(true);
             }
         }
         break;
 
 #ifdef ENABLE_WALLET
```

vasild commented on doc/release-notes-15936.md:
```diff
@@ -0,0 +1,8 @@
+proxy settings, UPNP preferences) are now saved to <datadir>/settings.json file
+rather than to the Qt settings backend (windows registry or unix desktop config
+files), and the GUI settings will now be used if bitcoind is started in
+subsequently, rather than ignored.
```
This can be marked as resolved. I cannot do that for some reason even though I opened it.

_Is not the author of a suggestion the best one to judge whether it is resolved or not, github `~!(@*(*~!=&/_*`_

vasild commented on src/qt/optionsmodel.cpp:
```diff
@@ -19,18 +19,93 @@
+    if (value.isNull()) return fallback;
+    if (value.isBool()) return value.get_bool();
+    if (value.isNum()) return value.get_int();
+    return LocaleIndependentAtoi<int>(value.get_str().c_str());
```
Can be marked as resolved.

vasild commented on src/qt/optionsmodel.cpp:
```diff
@@ -19,18 +19,93 @@
+{
+    if (value.isNull()) return fallback;
+    if (value.isFalse()) return "";
+    return QString::fromStdString(value.get_str());
```
Can be marked as resolved.

vasild commented on src/qt/optionsmodel.cpp:
```diff
@@ -19,18 +19,93 @@
+{
+    if (!variant.isValid()) return fallback;
+    if (type == QVariant::Bool) return variant.toBool();
+    if (type == QVariant::Int) return variant.toInt();
```
Can be marked as resolved.

vasild commented on src/qt/optionsmodel.cpp:
```diff
@@ -19,18 +19,93 @@
+    if (type == QVariant::Bool) return variant.toBool();
+    if (type == QVariant::Int) return variant.toInt();
+    std::string str = variant.toString().toStdString();
+    if (str.empty()) return false;
```
Can be marked as resolved.

vasild commented on src/qt/optionsmodel.h:
```diff
@@ -79,6 +79,8 @@ class OptionsModel : public QAbstractListModel
     QVariant data(const QModelIndex & index, int role = Qt::DisplayRole) const override;
     bool setData(const QModelIndex & index, const QVariant & value, int role = Qt::EditRole) override;
+    QVariant getOption(OptionID option) const;
+    bool setOption(OptionID option, const QVariant &value);
```
Can be marked as resolved.

vasild commented on src/qt/optionsmodel.cpp:
```diff
@@ -265,322 +266,337 @@ int OptionsModel::rowCount(const QModelIndex & parent) const
+    // this point because this function is only called after the intro screen is
+    // shown, before the node starts.
+    node().forceSetting("prune", new_value);
+    m_prune_size_gb = PruneSizeGB(new_value);
```
Can be marked as resolved.

vasild commented on src/util/system.cpp:
```diff
@@ -600,18 +607,33 @@ bool ArgsManager::IsArgNegated(const std::string& strArg) const
+    return SettingToBool(value, fDefault);
+}
+
+bool SettingToBool(const util::SettingsValue& value, bool fDefault)
+{
     return value.isNull() ? fDefault : value.isBool() ? value.get_bool() : InterpretBool(value.get_str());
 }
```
Can be marked as resolved.

vasild commented on src/qt/optionsmodel.cpp:
```diff
@@ -265,322 +225,342 @@ int OptionsModel::rowCount(const QModelIndex & parent) const
+        break;
+    case DatabaseCache:
+        if (changed()) {
+            node().updateSetting("dbcache", value.toInt());
```
Can be marked as resolved.

vasild commented on src/qt/optionsmodel.cpp:
```diff
@@ -637,4 +617,49 @@ void OptionsModel::checkAndMigrate()
+    migrate_setting(PruneSize, "nPruneSize", "prune");
+    migrate_setting(Prune, "bPrune", "prune");
+    migrate_setting(ProxyIP, "addrProxy", "proxy");
+    migrate_setting(ProxyUse, "fUseProxy", "proxy");
```
Alright, if this PR settles as it is (to store "use proxy" implicitly), then this is ok IMO. If it is decided to store "use proxy" explicitly then this will be changed anyway.

Can be marked as resolved.

vasild commented:
[Coverage report](https://people.freebsd.org/~vd/pr15936_99b4ce266f_coverage/modified_and_not_covered.html) for modified lines by this PR and not covered by tests (`test_bitcoin`, `test_bitcoin-qt`, functional tests).

ryanofsky reviewed (commented):
Thanks vasild for the more thorough reviews! I implmented all your suggestions, including the one that retains disabled proxy and prune setting across restarts, except that one is a separate commit not currently included in the PR because of the complexity it adds (see comments below).

Updated 99b4ce266f62afc8a37575527d82ac6048f50f75 -> 633596af152ddc893269787b227d2ea3f9c1c899 ([`pr/qtset.39`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.39) -> [`pr/qtset.40`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.40), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/qtset.39..pr/qtset.40)) with various suggestions and fixes from vasild, also fixing an fs::quoted windows build error https://cirrus-ci.com/task/5838702671822848

re: https://github.com/bitcoin/bitcoin/pull/15936#discussion_r852857701

Probably Hennadii could confirm, but I think the only reason the commit was called a prerequisite was that at the time, getOption() method did not exist, so OptionsModel didn't provide a convenient way to get the current UPNP value, while setting a new NATPMP value, or vice versa. So the mapPort call was moved out of OptionsModel to OptionsDialog, where the call didn't really belong, but both option values are easily available. However, OptionsModel always made more sense than for the call than OptionsDialog, because OptionsModel is the place where every other setting is applied (also the place that makes sense conceptually, for model/view separation).

vasild commented:
633596af152ddc893269787b227d2ea3f9c1c899 looks ok, modulo the [vanishing proxy values](https://github.com/bitcoin/bitcoin/pull/15936#discussion_r852998379), I have to think more about this. I will also test a bit further. Thanks!

ryanofsky reviewed (commented):
re: https://github.com/bitcoin/bitcoin/pull/15936#issuecomment-1110814375

> [633596a](https://github.com/bitcoin/bitcoin/commit/633596af152ddc893269787b227d2ea3f9c1c899) looks ok, modulo the [vanishing proxy values](https://github.com/bitcoin/bitcoin/pull/15936#discussion_r852998379), I have to think more about this. I will also test a bit further. Thanks!

Thanks again for the reviews! I would probably qualify them "vanishing disabled values after restarting" instead of "vanishing values" but I opened https://github.com/bitcoin-core/gui/issues/596 to think about this more and get more feedback

Sjors commented on src/qt/optionsmodel.cpp:
```diff
@@ -98,7 +104,17 @@ void OptionsModel::Init(bool resetSettings)
+        try {
+            getOption(option);
+        } catch (const std::runtime_error& e) {
+            error = strprintf(_("Could not read setting \"%s\", %s."), setting, e.what());
```
68a4f322e93e413cb8f1f1b998a2b2ee63df9b9d: how can you trigger this error?

vasild replied:
The only way to throw from `getOption()` is if `SettingName()` (called from `getOption()`) throws. However, `SettingName(option)` (same argument) is also called just before this `try/catch` without `try/catch`. So, if `SettingName()` is going to throw it will do so before this `try/catch` and so this code will never be reached.

To trigger that `SettingName()` to throw the source code must be modified. Cannot be triggered by wrong user input or corrupted config file.

Sjors reviewed (commented):
Did some code review up to 1e1ccabf0d38201e1c37164c7ec02ff8036af820.

Lightly tested some of the settings migrations. This should be thoroughly tested though, especially the Tor proxy, since a mistake there could have privacy consequences.

Once all settings are migrated, the following text needs to be adjusted: "Options set in this dialog are overridden by the command line ~or in the configuration file~". The way it seems to work now (which is fine by me) is that that `settings.json` takes precedent over `bitcoin.conf`, whereas previously `bitcoin.conf` would take precedence over GUI settings.

The above change has some subtle consequences, that are probably fine. If before this PR you set `dbcache=1000` in `bitcoin.conf` and `999` in the GUI settings, it would use `1000`. This PR  wipes `999` from QT's settings and sets the field to `1000`. This is definitely safer than the alternative of migrating the GUI setting and suddenly starting to use it. It may be good to point this behaviour out in the release notes.

ryanofsky reviewed (commented):
re: https://github.com/bitcoin/bitcoin/pull/15936#pullrequestreview-958013757

> Once all settings are migrated, the following text needs to be adjusted: "Options set in this dialog are overridden by the command line ~or in the configuration file~".

This change is made in the commit migrating the last setting which would be overridden by the configuration file (commit "Migrate -lang setting from QSettings to settings.json" af8ac293addda5778a52cc39bc270a0c9179bafb). I think this is after the commits you looked at.

> It may be good to point this behaviour out in the release notes.

Thanks, updated to mention the `bitcoin.conf` interaction. All the behavior you described is intentional. Applying this PR or reverting it should keep applied settings values the same.

---

Updated 633596af152ddc893269787b227d2ea3f9c1c899 -> 2f77c49174ca3932988f74faf75c8838dd95ec82 ([`pr/qtset.40`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.40) -> [`pr/qtset.41`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.41), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/qtset.40..pr/qtset.41)) adding comment and more release notes.

vasild commented:
While testing this, 22.x crashes if `settings.json` contains `"prune": 1234` due to https://github.com/bitcoin/bitcoin/pull/24498 which was fixed in 23.0. So, if this PR is included in 24.x and a user upgrades to 24.x and then downgrades to 22.x his 22.x would crash at startup. Is that acceptable?

ryanofsky commented:
> While testing this, 22.x crashes if `settings.json` contains `"prune": 1234` due to #24498 which was fixed in 23.0. So, if this PR is included in 24.x and a user upgrades to 24.x and then downgrades to 22.x his 22.x would crash at startup. Is that acceptable?

What a mess. I guess I'll just write a string instead of an int to avoid this. Maybe will try to backport #24498, too, since it's a one-line fix and would make 22.x less fragile in general. Thanks for testing this! It is pretty important for this PR to be backwards compatible.

ryanofsky committed:
settings: Add resetSettings() method

Allows the GUI to clear settings.json file and save settings.json.bak file when
GUI "Reset Options" button is pressed or -resetguisettings command line option
is used. (GUI code already backs up and resets the "guisettings.ini" file this
way, so this just makes the same behavior possible for "settings.json")

ryanofsky commented:
re: https://github.com/bitcoin/bitcoin/pull/15936#issuecomment-1116020233

> What a mess. I guess I'll just write a string instead of an int to avoid this

Updated so this now writes 22.x int settings as strings to avoid triggering uncaught exceptions if there is a downgrade. I also added a new commit to clear settings.json file when the reset button is pressed, after @Rspigler's comment https://github.com/bitcoin-core/gui/issues/596#issuecomment-1115233031

Updated 2f77c49174ca3932988f74faf75c8838dd95ec82 -> 616b95a061ae4b9de171277da1e97d3d6543b0b0 ([`pr/qtset.41`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.41) -> [`pr/qtset.42`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.42), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/qtset.41..pr/qtset.42)) with 22.x compatibility and reset button fixes

ryanofsky commented:
Updated 616b95a061ae4b9de171277da1e97d3d6543b0b0 -> e4ef0b6f78cc7c6ac148674b9c8ac40bd7e5f30d ([`pr/qtset.42`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.42) -> [`pr/qtset.43`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.43), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/qtset.42..pr/qtset.43)) to settings migrate test https://cirrus-ci.com/task/6271609857114112?logs=ci#L2726, https://cirrus-ci.com/task/5708659903692800?logs=ci#L3089, https://cirrus-ci.com/task/6553084833824768?logs=ci#L3820, https://cirrus-ci.com/task/5990134880403456?logs=ci#L2954, https://cirrus-ci.com/task/5849397392048128?logs=ci#L3419, https://cirrus-ci.com/task/5286447438626816?logs=ci#L2913, https://cirrus-ci.com/task/5004972461916160?logs=ci#L2304

ryanofsky commented:
This PR needs more ACKs, but it has been reviewed extensively and tested to death. I'd really appreciate any new review and testing, but also updates from people who commented previously. Summary of feedback so far:

- _jonasschnelli_ concept ACK https://github.com/bitcoin/bitcoin/pull/15936#issuecomment-493666091
- _marcofalke_ review https://github.com/bitcoin/bitcoin/pull/15936#pullrequestreview-770498004
- _prayank23_ ACK https://github.com/bitcoin/bitcoin/pull/15936#pullrequestreview-775523650 and testing https://github.com/bitcoin/bitcoin/pull/15936#issuecomment-938400093
- _hebasto_ approach ACK https://github.com/bitcoin/bitcoin/pull/15936#pullrequestreview-852203064 and testing https://github.com/bitcoin/bitcoin/pull/15936#pullrequestreview-852727704
- _rspigler_ testing https://github.com/bitcoin/bitcoin/pull/15936#issuecomment-1012773648
- _jonatack_ concept ACK https://github.com/bitcoin/bitcoin/pull/15936#issuecomment-1076312880
- _vasild_ numerous reviews, closest to ACK https://github.com/bitcoin/bitcoin/pull/15936#issuecomment-1110814375
- _sjors_ partial review https://github.com/bitcoin/bitcoin/pull/15936#pullrequestreview-958013757

ryanofsky commented:
Rebased e4ef0b6f78cc7c6ac148674b9c8ac40bd7e5f30d -> 084b888ef1934f4ef88da33f72ffdd321af7c6b7 ([`pr/qtset.43`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.43) -> [`pr/qtset.44`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.44), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/qtset.43-rebase..pr/qtset.44)) due to conflict with #25153

ryanofsky commented:
Updated 084b888ef1934f4ef88da33f72ffdd321af7c6b7 -> 02ce0badbe3fd50a508b1c189082c91bff420c60 ([`pr/qtset.44`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.44) -> [`pr/qtset.45`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.45), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/qtset.44..pr/qtset.45)) dropping the GUI changes from this PR and moving them to https://github.com/bitcoin-core/gui/pull/602. This should make the node changes here, and the GUI changes over there both easier to review

ryanofsky commented:
Updated 084b888ef1934f4ef88da33f72ffdd321af7c6b7 -> 02ce0badbe3fd50a508b1c189082c91bff420c60 ([`pr/qtset.44`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.44) -> [`pr/qtset.45`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.45), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/qtset.44..pr/qtset.45)) dropping the GUI changes from this PR and moving them to https://github.com/bitcoin-core/gui/pull/602. This should make the node changes here, and the GUI changes over there both easier to review

hebasto commented on src/util/system.cpp:
```diff
@@ -572,10 +575,10 @@ bool ArgsManager::ReadSettingsFile(std::vector<std::string>* errors)
 {
     fs::path path, path_tmp;
-    if (!GetSettingsPath(&path, /* temp= */ false) || !GetSettingsPath(&path_tmp, /* temp= */ true)) {
+    if (!GetSettingsPath(&path, /* temp= */ false, backup) || !GetSettingsPath(&path_tmp, /* temp= */ true, backup)) {
```
02ce0badbe3fd50a508b1c189082c91bff420c60, style nit

```suggestion
    if (!GetSettingsPath(&path, /*temp=*/false, backup) || !GetSettingsPath(&path_tmp, /*temp=*/true, backup)) {
```

hebasto reviewed (approved):
ACK 02ce0badbe3fd50a508b1c189082c91bff420c60, I have reviewed the code and it looks OK, I agree it can be merged.

Thanks for 9dd3817ee2f67d021aad7a809daa136300a8d8e1 _"init: Remove Shutdown() node.args reset"_ commit!

Only style nits, including following parameter naming convention in new `SettingTo{String,Int,Bool}` functions. Anyway, feel free to ignore :)

ryanofsky commented:
Thanks for the review!

Updated 02ce0badbe3fd50a508b1c189082c91bff420c60 -> 3340286c752d2b8ace594f7ae179c290b3687a37 ([`pr/qtset.45`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.45) -> [`pr/qtset.46`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.46), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/qtset.45..pr/qtset.46)) just tweaking some comments, including the parameter name one that was suggested.

hebasto reviewed (approved):
re-ACK 3340286c752d2b8ace594f7ae179c290b3687a37

Rspigler commented:
I will re-test tomorrow.  Thanks for keeping this up!

vasild reviewed (approved):
ACK 3340286c752d2b8ace594f7ae179c290b3687a37

Rspigler commented:
tACK 3340286c752d2b8ace594f7ae179c290b3687a37

Tested with various changes to `settings.json`, `bitcoin.conf`, and the GUI.  Everything behaved as it should.

Previous issues people reported are resolved (including https://github.com/bitcoin/bitcoin/issues/24457).

Only other issue is solved by https://github.com/bitcoin-core/gui/pull/603

promag commented on src/node/interfaces.cpp:
```diff
@@ -112,6 +112,38 @@ class NodeImpl : public Node
+    util::SettingsValue getPersistentSetting(const std::string& name) override { return gArgs.GetPersistentSetting(name); }
+    void updateSetting(const std::string& name, const util::SettingsValue& value) override
+    {
+        gArgs.LockSettings([&](util::Settings& settings) {
```
```cpp
updateRwSetting(name, value, /* write = */ true)
```

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/15936#discussion_r878685831

> ```c++
> updateRwSetting(name, value, /* write = */ true)
> ```

Thanks renamed to `updateRwSetting` to be consistent with the Chain interface

promag reviewed (commented):
Concept ACK, unifying persistent node settings across apps makes sense to me.

Rspigler commented:
tACK https://github.com/bitcoin/bitcoin/commit/3340286c752d2b8ace594f7ae179c290b3687a37 all tests pass

ryanofsky reviewed (commented):
Updated 3340286c752d2b8ace594f7ae179c290b3687a37 -> f9fdcec7e932843a91ddf7f377e00bd2a6efb82a ([`pr/qtset.46`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.46) -> [`pr/qtset.47`](https://github.com/ryanofsky/bitcoin/commits/pr/qtset.47), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/qtset.46..pr/qtset.47)) just renaming `updateSetting` to `updateRwSetting`;

vasild reviewed (approved):
ACK f9fdcec7e932843a91ddf7f377e00bd2a6efb82a

hebasto reviewed (approved):
re-ACK f9fdcec7e932843a91ddf7f377e00bd2a6efb82a, only a function renamed since my recent [review](https://github.com/bitcoin/bitcoin/pull/15936#pullrequestreview-979324357).

ryanofsky reviewed (commented):
This is probably ready for merge (sorry for the confusing history, it was originally a lot bigger PR before I removed the GUI parts)
